File: BTC5/C++/Trie.cpp
#include <bits/stdc++.h>

using namespace std;

struct trie {
    vector<trie *> child;
    bool hasWord;

    trie() 
	{
        child.resize(26, 0);
        hasWord = 0;
    }
};

void add(trie *t, string s) 
{
    for (int i = 0; i < s.size(); ++i)	
	{
		if (!t->child[s[i] - 'a'])	t->child[s[i] - 'a'] = new trie();
		t = t->child[s[i] - 'a'];
	}
	t->hasWord = 1;
}

bool haveOnlyChild(trie * root) 
{
	int cnt = 0;
	for (int i = 0; i < 26; ++i)	if (root -> child[i])	++cnt;
	return cnt == 1;
}

string findLCP(trie * root) 
{
	string LCP = "";
	while (root && haveOnlyChild(root))
	{
		for (int i = 0; i < 26; ++i) 
		{
			if (root -> child[i]) 
			{
				root = root -> child[i];
				LCP += (char) (i + 'a');
				break;
			}
		}	
	}
	return LCP;
}
int main() {
    trie *t = new trie();
    vector<string> words = vector<string>{"flower", "flow", "floght", "flot"};

    for (string temp : words)	add(t, temp);

	cout << findLCP(t);
    return 0;
}

File: BTC5/Python/Trie.py
class TrieNode:
    def __init__(self):
        self.hasWord = False
        self.char = {}

def insert(root, s):
    for c in s:
        root = root.char.setdefault(c, TrieNode())
    root.hasWord = True

def findLCP(root):
    lcp = ''
    while root and len(root.char) == 1:
        for key, val in root.char.items():
            lcp += key
            root = val
    return lcp or None

if __name__ == '__main__':
    root = TrieNode()
    words = ["flower", "flow", "floght"]
    for s in words:
        insert(root, s)
    print('Tien to chung dai nhat:', findLCP(root))